flatland_baselines/deadlock_avoidance_heuristic/policy/deadlock_avoidance_policy.py

```diff
@@ -188,18 +189,38 @@ def _extract_agent_can_move(self):
             plt.show(block=False)
             plt.pause(0.01)
 
-    def _check_agent_can_move(self,
-                              handle,
-                              my_shortest_walking_path,
-                              same_agents,
-                              opp_agents,
-                              full_shortest_distance_agent_map):
-        agent_positions_map = (self.agent_positions > -1).astype(int)
+    def _check_agent_can_move(
+            self,
+            my_shortest_walking_path: np.ndarray,
+            opp_agents: Set,
+            full_shortest_distance_agent_map: np.ndarray,
+            switches
+    ):
+        """
+        The algorithm collects for each train along its route all trains that are currently on a resource in the route.
+        For each collected train the method has to decide whether a deadlock between the train and the collected train occurs.
+        If no deadlock is found, the process must further decide at which position along the route the train must let pass the collected train.
+        This can be achieved by searching the train's path required resources backward along the path starting at the collected train position.
+        Stop the search when the resource along the collected train's path is not equal.
+        The forward and backward traveling along the train and the collected train path must be done step-by-step synchronous.
+        If the first non-equal resource position along the train's path is more than one resource from train's current location away,
+        then the train can move and no deadlock will occur for the next time step.
+
+        Parameters
+        ----------
+        my_shortest_walking_path : np.ndarray
+        opp_agents : np.ndarray
+        full_shortest_distance_agent_map : np.ndarray
+
+        Returns
+        -------
+
+        """
+
         len_opp_agents = len(opp_agents)
         for opp_a in opp_agents:
             opp = full_shortest_distance_agent_map[opp_a]
-            delta = ((my_shortest_walking_path - opp - agent_positions_map) > 0).astype(int)
-            sum_delta = np.sum(delta)
+            sum_delta = np.count_nonzero((my_shortest_walking_path - switches - opp) > 0)
             if sum_delta < (self.min_free_cell + len_opp_agents):
```

Comment:
@aiAdrian can you explain the `+ len_opp_agents` term? By construction `my_shortest_walking_path` is the path up to the first oncoming agent (non-including) and hence `sum_delta` does not include any of the oncoming agents, so why is this condition correct?